File: 2015/06/solution.py
import re
import sys

import numpy

grid1 = numpy.zeros(shape=(1000, 1000), dtype="int8")
grid2 = grid1.copy()

for line in sys.stdin:
    r1, c1, r2, c2 = map(int, re.findall(r"\d+", line))
    r2 += 1
    c2 += 1
    op =line.split()[1]
    if op == "on":
        grid1[r1:r2, c1:c2] = 1
        grid2[r1:r2, c1:c2] += 1
    elif op == "off":
        grid1[r1:r2, c1:c2] = 0
        positive = grid2[r1:r2, c1:c2] > 0
        grid2[r1:r2, c1:c2][positive] -= 1
    else:
        zeros = grid1[r1:r2, c1:c2] == 0
        ones = grid1[r1:r2, c1:c2] == 1
        grid1[r1:r2, c1:c2][zeros] = 1
        grid1[r1:r2, c1:c2][ones] = 0
        grid2[r1:r2, c1:c2] += 2

print("Part One:", grid1.sum())
print("Part Two:", grid2.sum())

File: 2015/06/solution.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strings"
)

func main() {
	var grid1 [1000][1000]int
	var grid2 [1000][1000]int
	scanner := bufio.NewScanner(os.Stdin)
	for scanner.Scan() {
		line := scanner.Text()
		fields := strings.Fields(line)
		var cmd, start, end string
		if len(fields) == 4 {
			cmd, start, _, end = fields[0], fields[1], fields[2], fields[3]
		} else {
			_, cmd, start, _, end = fields[0], fields[1], fields[2], fields[3], fields[4]
		}
		var r1, c1, r2, c2 int
		fmt.Sscanf(start, "%d,%d", &r1, &c1)
		fmt.Sscanf(end, "%d,%d", &r2, &c2)
		for r := r1; r <= r2; r++ {
			for c := c1; c <= c2; c++ {
				switch cmd {
				case "toggle":
					if grid1[r][c] == 1 {
						grid1[r][c] = 0
					} else {
						grid1[r][c] = 1
					}
					grid2[r][c] += 2
				case "on":
					grid1[r][c] = 1
					grid2[r][c] += 1
				case "off":
					grid1[r][c] = 0
					if grid2[r][c] > 0 {
						grid2[r][c] -= 1
					}
				}
			}
		}
	}
	p1, p2 := 0, 0
	for r := range grid1 {
		for c := range grid1[r] {
			p1 += grid1[r][c]
			p2 += grid2[r][c]
		}
	}
	fmt.Println("Part One:", p1)
	fmt.Println("Part Two:", p2)
}
